src/typechecker.erl

```diff
@@ -4165,87 +4240,95 @@ add_type_pat(Pat, ?type(union, UnionTys)=UnionTy, Env, VEnv) ->
+            NewEnv = union_var_binds([ add_any_types_pat(Pat, Env) || Pat <- Pats ], Env),
+            {type(none),
+             Ty,
+             NewEnv,
+             constraints:empty()};
         {elem_ty, Tys, Cs} ->
-            {PatTys, UBounds, VEnv1, Cs1} = do_add_types_pats(Pats, Tys, Env, VEnv),
-            {type(tuple, PatTys)
-            ,type(tuple, UBounds)
-            ,VEnv1
-            ,constraints:combine(Cs, Cs1)};
+            {PatTys, UBounds, Env1, Cs1} = do_add_types_pats(Pats, Tys, Env),
+            {type(tuple, PatTys),
+             type(tuple, UBounds),
+             Env1,
+             constraints:combine(Cs, Cs1)};
         {type_error, _Type} ->
             throw({type_error, pattern, P, Tuple, Ty})
     end;
-add_type_pat(Atom = {atom, P, Val}, Ty, Env, VEnv) ->
+add_type_pat(Atom = {atom, P, Val}, Ty, Env) ->
     LitTy = {atom, erl_anno:new(0), Val},
     case subtype(LitTy, Ty, Env) of
         {true, Cs} ->
-            {LitTy, LitTy, VEnv, Cs};
+            {LitTy, LitTy, Env, Cs};
         false ->
             throw({type_error, pattern, P, Atom, Ty})
     end;
-add_type_pat(Nil = {nil, P}, Ty, Env, VEnv) ->
+add_type_pat(Nil = {nil, P}, Ty, Env) ->
     NilTy = type(nil),
     case subtype(NilTy, Ty, Env) of
         {true, Cs} ->
-            {NilTy, NilTy, VEnv, Cs};
+            {NilTy, NilTy, Env, Cs};
         false ->
             throw({type_error, pattern, P, Nil, Ty})
     end;
-add_type_pat(CONS = {cons, P, PH, PT}, ListTy, Env, VEnv) ->
+add_type_pat(CONS = {cons, P, PH, PT}, ListTy, Env) ->
     case expect_list_type(normalize(ListTy, Env), dont_allow_nil_type, Env) of
         any ->
-            VEnv2 = add_any_types_pat(PH, VEnv),
+            Env2 = add_any_types_pat(PH, Env),
             TailTy = normalize(type(union, [ListTy, type(nil)]), Env),
-            {_TailPatTy, _TauUBound, VEnv3, Cs} = add_type_pat(PT, TailTy, Env, VEnv2),
+            {_TailPatTy, _TauUBound, VEnv3, Cs} = add_type_pat(PT, TailTy, Env2),
             NonEmptyTy = rewrite_list_to_nonempty_list(ListTy),
-            {NonEmptyTy, NonEmptyTy, VEnv3, Cs};
+            {type(none), NonEmptyTy, VEnv3, Cs};
         {elem_ty, ElemTy, Cs1} ->
-            {_PatTy1, _UBound1, VEnv2, Cs2} =
-                add_type_pat(PH, normalize(ElemTy, Env), Env, VEnv),
+            {PatTy1, _UBound1, Env2, Cs2} =
+                add_type_pat(PH, normalize(ElemTy, Env), Env),
             TailTy = normalize(type(union, [ListTy, type(nil)]), Env),
-            {_PatTy2, _Ubound2, VEnv3, Cs3} = add_type_pat(PT, TailTy, Env, VEnv2),
+            {_PatTy2, _Ubound2, Env3, Cs3} = add_type_pat(PT, TailTy, Env2),
+            {PatTy, Env4} = case is_list_pat_exhaustive(CONS) of
+                                true ->
+                                    {type(nonempty_list, [PatTy1]), Env3};
+                                false ->
+                                    {type(none), disable_exhaustiveness_check(Env3)}
```

Comment:
Do we really always need to disable exhaustiveness-checking when a pattern is nonexhaustive? For example we could give an error for f below, but for g we can't know.

```Erlang
-spec f(string()) -> ok;
f([$a]) -> ok;
f([$a | _]) -> ok.

-spec g(string()) -> ok;
g([]) -> ok;
g([_] -> ok;
g([_, _ | _]) -> ok.
```

Reply:
There's definitely room for improvement here, but I'd prefer going in small steps, i.e. in subsequent PRs.

Reply:
This is now tracked as #409.